ui-lib/src/main/java/co/electriccoin/zcash/work/SyncWorker.kt

```diff
@@ -12,55 +12,115 @@ import cash.z.ecc.android.sdk.Synchronizer
+import kotlinx.datetime.Clock
+import kotlinx.datetime.DateTimeUnit
+import kotlinx.datetime.Instant
+import kotlinx.datetime.TimeZone
+import kotlinx.datetime.atTime
+import kotlinx.datetime.toInstant
+import kotlinx.datetime.toLocalDateTime
+import kotlinx.datetime.until
+import kotlin.random.Random
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.days
 import kotlin.time.Duration.Companion.hours
+import kotlin.time.Duration.Companion.minutes
+import kotlin.time.DurationUnit
+import kotlin.time.toDuration
 import kotlin.time.toJavaDuration
 
+// TODO [#1249]: Add documentation and tests on background syncing
+// TODO [#1249]: https://github.com/Electric-Coin-Company/zashi-android/issues/1249
+
 @Keep
 class SyncWorker(context: Context, workerParameters: WorkerParameters) : CoroutineWorker(context, workerParameters) {
     @OptIn(ExperimentalCoroutinesApi::class)
     override suspend fun doWork(): Result {
-        // Enhancements to this implementation would be:
-        // - Quit early if the synchronizer is null after a timeout period
-        // - Return better status information
+        Twig.debug { "BG Sync: starting..." }
 
         WalletCoordinator.getInstance(applicationContext).synchronizer
             .flatMapLatest {
+                Twig.debug { "BG Sync: synchronizer: $it" }
+
                 it?.status?.combine(it.progress) { status, progress ->
-                    StatusAndProgress(status, progress)
+                    StatusAndProgress(status, progress).also {
+                        Twig.debug { "BG Sync: result: $it" }
+                    }
                 } ?: emptyFlow()
             }
             .takeWhile {
-                it.status != Synchronizer.Status.DISCONNECTED && it.progress.isLessThanHundredPercent()
+                it.status != Synchronizer.Status.DISCONNECTED &&
+                    it.status != Synchronizer.Status.SYNCED
             }
             .collect()
 
+        Twig.debug { "BG Sync: terminating..." }
+
         return Result.success()
     }
 
     companion object {
-        /*
-         * There may be better periods; we have not optimized for this yet.
-         */
-        private val DEFAULT_SYNC_PERIOD = 24.hours
+        private val SYNC_PERIOD = 24.hours
+        private val SYNC_DAY_SHIFT = 1.days // Move to tomorrow
+        private val SYNC_START_TIME_HOURS = 3.hours // Start around 3 a.m. at night
+        private val SYNC_START_TIME_MINUTES = 60.minutes // Randomize with minutes until 4 a.m.
 
         fun newWorkRequest(): PeriodicWorkRequest {
+            val targetTimeDiff = calculateTargetTimeDifference()
+
+            Twig.debug { "BG Sync: necessary trigger delay time: $targetTimeDiff" }
+
             val constraints =
                 Constraints.Builder()
                     .setRequiresStorageNotLow(true)
-                    .setRequiredNetworkType(NetworkType.CONNECTED)
+                    .setRequiredNetworkType(NetworkType.UNMETERED)
```

Comment:
One question; since this is a change to an existing method, doesn't this mean that this now makes it that sync can _only_ happen on wifi and while charging?

Reply:
The existing solution of background syncing wasn't working as expected (see #634) and had a problem with its triggering because of the `it.progress.isLessThanHundredPercent()` condition, although it was enabled for users. These issues are now fixed.